File: .github/scripts/validate-extended-profile-contracts.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative "lib/profile_contracts"

include ProfileContracts

begin
  skill = SkillDocument.read("SKILL.md")
  selection = ProfileSelection.load("SKILL.md", document: skill)
rescue ParseError => e
  warn e.message
  exit 1
end

repository = RepositorySnapshot.new
selected_profiles = selection.profiles
errors = []

if selection.template_scaffold?
  scaffold_directories = %w[
    references assets scripts mcp src app lib bin server client tests
    web website frontend ui public static www
  ]
  customized = scaffold_directories.select { |directory| repository.operational_file_present?(directory) }
  manifests = %w[
    package.json package-lock.json pnpm-lock.yaml yarn.lock bun.lock bun.lockb
    pyproject.toml requirements.txt uv.lock Pipfile Pipfile.lock Cargo.toml Cargo.lock
    go.mod go.sum Gemfile Gemfile.lock pom.xml build.gradle build.gradle.kts
    composer.json composer.lock
  ].select { |path| repository.file?(path) }
  roots = %w[
    index.html service-worker.js sw.js manifest.webmanifest Dockerfile compose.yml
    compose.yaml docker-compose.yml docker-compose.yaml
  ].select { |path| repository.file?(path) }

  errors << "'template-scaffold' cannot be retained after adding implementation or operational files under: #{customized.join(', ')}." unless customized.empty?
  errors << "'template-scaffold' cannot be retained after adding runtime or package manifests: #{manifests.join(', ')}." unless manifests.empty?
  errors << "'template-scaffold' cannot be retained after adding root implementation or deployment files: #{roots.join(', ')}." unless roots.empty?
  unless skill.metadata["name"] == "agent-skill-template"
    errors << "'template-scaffold' is valid only while the skill name remains 'agent-skill-template'."
  end
end

skill.declarations("Script").each do |declaration|
  next if declaration.path == "scripts/TODO"

  required_fields = [
    "Run when",
    "Exact invocation",
    "Working directory",
    "Inputs and arguments",
    "Stdout/result",
    "Stderr/diagnostics",
    "Exit status",
    "Files or external state modified",
    "Network access",
    "Required permissions",
    "Automatic execution allowed",
    "Human confirmation required",
    "Idempotency and retry behavior"
  ]
  allow_none = [
    "Inputs and arguments",
    "Files or external state modified",
    "Network access",
    "Required permissions"
  ]

  required_fields.each do |field|
    value = declaration.fields[field]
    valid = allow_none.include?(field) ? ValuePolicy.resolved?(value) : ValuePolicy.concrete?(value)
    unless valid
      errors << "SKILL.md script declaration for #{declaration.path} must include a concrete '#{field}:' value."
    end
  end

  {
    "Automatic execution allowed" => declaration.fields["Automatic execution allowed"],
    "Human confirmation required" => declaration.fields["Human confirmation required"]
  }.each do |field, value|
    if ValuePolicy.resolved?(value) && !["YES", "NO", "WITH CONDITIONS"].include?(value.upcase)
      errors << "SKILL.md script declaration for #{declaration.path} must set '#{field}:' to YES, NO, or WITH CONDITIONS."
    end
  end
end

runtime_profiles = %w[script-assisted packaged-cli mcp-enabled browser-interface headless-service]
runtime = repository.document("RUNTIME.md")
if (selected_profiles & runtime_profiles).any? && runtime
  primary = runtime.section("## Primary implementation")
  ["Language", "Runtime", "Minimum runtime version", "Source layout", "Supported operating systems"].each do |item|
    unless ValuePolicy.concrete?(runtime.table_value(item, section: primary))
      errors << "Selected runtime-backed profiles require a concrete '#{item}' value in RUNTIME.md."
    end
  end
end

if selection.selected?("browser-interface")
  web = repository.document("WEB_INTERFACE.md")
  if web
    relationship = web.section("## Relationship to MCP")
    models = [
      "backend acts as an MCP client",
      "browser calls MCP directly",
      "UI uses a non-MCP application API",
      "mixed model"
    ].to_h { |label| [label, web.list_field(label, section: relationship)&.upcase] }

    unless models.values.all? { |value| %w[YES NO].include?(value) }
      errors << "WEB_INTERFACE.md must set every UI interaction model to YES or NO."
    end
    unless models.values.count("YES") == 1
      errors << "WEB_INTERFACE.md must select exactly one UI interaction model with YES."
    end
  end
end

if selection.selected?("mcp-enabled")
  mcp = repository.document("MCP_INTERFACE.md")
  runtime = repository.document("RUNTIME.md")
  if mcp
    variants = {
      "stdio" => {
        heading: "## stdio MCP server variant",
        runtime_heading: "### stdio variant",
        mandatory: ["Launch command", "Lifecycle owner"]
      },
      "Streamable HTTP" => {
        heading: "## Streamable HTTP MCP server variant",
        runtime_heading: "### Streamable HTTP variant",
        mandatory: [
          "Start command",
          "Stop command or shutdown method",
          "Endpoint URL",
          "Bind address",
          "Port selection",
          "Supported protocol eras",
          "Revision-specific state model",
          "Authentication",
          "Health/readiness check"
        ]
      },
      "bundled MCP client" => {
        heading: "## Bundled ad hoc MCP tool client",
        runtime_heading: "### Bundled ad hoc MCP tool client",
        mandatory: [
          "Scope",
          "Command",
          "Transport used",
          "Negotiation and compatibility behavior",
          "Invocation scope",
          "Interaction modes",
          "Task or extension support"
        ],
        allow_not_supported: ["Task or extension support"]
      }
    }

    supported_server_variants = []
    variants.each do |variant_name, spec|
      section = mcp.section(spec[:heading])
      if section.nil?
        errors << "Selected profile 'mcp-enabled' requires '#{spec[:heading]}' in MCP_INTERFACE.md."
        next
      end

      interface_support = ProfileContracts.support_token(mcp.field("Supported", section: section))
      unless %w[YES NO].include?(interface_support)
        errors << "MCP interface '#{variant_name}' must set 'Supported:' to YES or NO in MCP_INTERFACE.md."
        next
      end

      if runtime
        runtime_section = runtime.section(spec[:runtime_heading])
        runtime_support = ProfileContracts.support_token(runtime.table_value("Supported", section: runtime_section))
        unless %w[YES NO].include?(runtime_support)
          errors << "MCP variant '#{variant_name}' must set Supported to YES or NO in RUNTIME.md."
        end
        if %w[YES NO].include?(runtime_support) && interface_support != runtime_support
          errors << "MCP variant '#{variant_name}' has inconsistent Supported values between RUNTIME.md and MCP_INTERFACE.md."
        end
      end

      next unless interface_support == "YES"

      supported_server_variants << variant_name unless variant_name == "bundled MCP client"
      if /\b(?:TODO|UNSELECTED)\b/i.match?(section)
        errors << "Supported MCP interface '#{variant_name}' must not retain TODO or UNSELECTED fields in MCP_INTERFACE.md."
      end

      spec[:mandatory].each do |label|
        value = mcp.field(label, section: section)
        valid = if Array(spec[:allow_not_supported]).include?(label)
                  ValuePolicy.resolved?(value)
                else
                  ValuePolicy.concrete?(value)
                end
        unless valid
          errors << "Supported MCP interface '#{variant_name}' requires a concrete '#{label}:' value in MCP_INTERFACE.md."
        end
      end
    end

    if supported_server_variants.empty?
      errors << "Selected profile 'mcp-enabled' requires at least one supported MCP server variant in MCP_INTERFACE.md."
    end
  end
end

unless errors.empty?
  errors.uniq.each { |error| warn error }
  exit 1
end

puts "Extended Agent Skill profile contracts are valid."

File: template/.github/scripts/validate_extended_profile_contracts.py
#!/usr/bin/env python3
"""Validate extended Agent Skill profile contracts."""

from __future__ import annotations

import re
import sys

from lib.profile_contracts import (
    ParseError,
    ProfileSelection,
    RepositorySnapshot,
    SkillDocument,
    ValuePolicy,
    support_token,
)


def run() -> int:
    try:
        skill = SkillDocument.read("SKILL.md")
        selection = ProfileSelection.load("SKILL.md", document=skill)
    except (ParseError, OSError) as exc:
        print(exc, file=sys.stderr)
        return 1

    repository = RepositorySnapshot()
    selected_profiles = list(selection.profiles)
    errors: list[str] = []

    if selection.template_scaffold():
        scaffold_directories = [
            "references",
            "assets",
            "scripts",
            "mcp",
            "src",
            "app",
            "lib",
            "bin",
            "server",
            "client",
            "tests",
            "web",
            "website",
            "frontend",
            "ui",
            "public",
            "static",
            "www",
        ]
        customized = [
            directory
            for directory in scaffold_directories
            if repository.operational_file_present(directory)
        ]
        manifests = [
            path
            for path in (
                "package.json",
                "package-lock.json",
                "pnpm-lock.yaml",
                "yarn.lock",
                "bun.lock",
                "bun.lockb",
                "pyproject.toml",
                "requirements.txt",
                "uv.lock",
                "Pipfile",
                "Pipfile.lock",
                "Cargo.toml",
                "Cargo.lock",
                "go.mod",
                "go.sum",
                "Gemfile",
                "Gemfile.lock",
                "pom.xml",
                "build.gradle",
                "build.gradle.kts",
                "composer.json",
                "composer.lock",
            )
            if repository.file(path)
        ]
        roots = [
            path
            for path in (
                "index.html",
                "service-worker.js",
                "sw.js",
                "manifest.webmanifest",
                "Dockerfile",
                "compose.yml",
                "compose.yaml",
                "docker-compose.yml",
                "docker-compose.yaml",
            )
            if repository.file(path)
        ]

        if customized:
            errors.append(
                "'template-scaffold' cannot be retained after adding "
                "implementation or operational files under: "
                + ", ".join(customized)
                + "."
            )
        if manifests:
            errors.append(
                "'template-scaffold' cannot be retained after adding runtime or "
                "package manifests: "
                + ", ".join(manifests)
                + "."
            )
        if roots:
            errors.append(
                "'template-scaffold' cannot be retained after adding root "
                "implementation or deployment files: "
                + ", ".join(roots)
                + "."
            )
        if skill.metadata.get("name") != "agent-skill-template":
            errors.append(
                "'template-scaffold' is valid only while the skill name remains "
                "'agent-skill-template'."
            )

    for declaration in skill.declarations("Script"):
        if declaration.path == "scripts/TODO":
            continue

        required_fields = [
            "Run when",
            "Exact invocation",
            "Working directory",
            "Inputs and arguments",
            "Stdout/result",
            "Stderr/diagnostics",
            "Exit status",
            "Files or external state modified",
            "Network access",
            "Required permissions",
            "Automatic execution allowed",
            "Human confirmation required",
            "Idempotency and retry behavior",
        ]
        allow_none = {
            "Inputs and arguments",
            "Files or external state modified",
            "Network access",
            "Required permissions",
        }

        for field in required_fields:
            value = declaration.fields.get(field)
            valid = (
                ValuePolicy.resolved(value)
                if field in allow_none
                else ValuePolicy.concrete(value)
            )
            if not valid:
                errors.append(
                    f"SKILL.md script declaration for {declaration.path} must "
                    f"include a concrete '{field}:' value."
                )

        for field in (
            "Automatic execution allowed",
            "Human confirmation required",
        ):
            value = declaration.fields.get(field)
            if (
                ValuePolicy.resolved(value)
                and str(value).upper()
                not in {"YES", "NO", "WITH CONDITIONS"}
            ):
                errors.append(
                    f"SKILL.md script declaration for {declaration.path} must "
                    f"set '{field}:' to YES, NO, or WITH CONDITIONS."
                )

    runtime_profiles = {
        "script-assisted",
        "packaged-cli",
        "mcp-enabled",
        "browser-interface",
        "headless-service",
    }
    runtime = repository.document("RUNTIME.md")
    if set(selected_profiles) & runtime_profiles and runtime is not None:
        primary = runtime.section("## Primary implementation")
        for item in (
            "Language",
            "Runtime",
            "Minimum runtime version",
            "Source layout",
            "Supported operating systems",
        ):
            if not ValuePolicy.concrete(
                runtime.table_value(item, section=primary)
            ):
                errors.append(
                    "Selected runtime-backed profiles require a concrete "
                    f"'{item}' value in RUNTIME.md."
                )

    if selection.selected("browser-interface"):
        web = repository.document("WEB_INTERFACE.md")
        if web is not None:
            relationship = web.section("## Relationship to MCP")
            models = {
                label: (
                    web.list_field(label, section=relationship).upper()
                    if web.list_field(label, section=relationship) is not None
                    else None
                )
                for label in (
                    "backend acts as an MCP client",
                    "browser calls MCP directly",
                    "UI uses a non-MCP application API",
                    "mixed model",
                )
            }

            if not all(value in {"YES", "NO"} for value in models.values()):
                errors.append(
                    "WEB_INTERFACE.md must set every UI interaction model to "
                    "YES or NO."
                )
            if list(models.values()).count("YES") != 1:
                errors.append(
                    "WEB_INTERFACE.md must select exactly one UI interaction "
                    "model with YES."
                )

    if selection.selected("mcp-enabled"):
        mcp = repository.document("MCP_INTERFACE.md")
        runtime = repository.document("RUNTIME.md")
        if mcp is not None:
            variants = {
                "stdio": {
                    "heading": "## stdio MCP server variant",
                    "runtime_heading": "### stdio variant",
                    "mandatory": ["Launch command", "Lifecycle owner"],
                    "allow_not_supported": [],
                },
                "Streamable HTTP": {
                    "heading": "## Streamable HTTP MCP server variant",
                    "runtime_heading": "### Streamable HTTP variant",
                    "mandatory": [
                        "Start command",
                        "Stop command or shutdown method",
                        "Endpoint URL",
                        "Bind address",
                        "Port selection",
                        "Supported protocol eras",
                        "Revision-specific state model",
                        "Authentication",
                        "Health/readiness check",
                    ],
                    "allow_not_supported": [],
                },
                "bundled MCP client": {
                    "heading": "## Bundled ad hoc MCP tool client",
                    "runtime_heading": "### Bundled ad hoc MCP tool client",
                    "mandatory": [
                        "Scope",
                        "Command",
                        "Transport used",
                        "Negotiation and compatibility behavior",
                        "Invocation scope",
                        "Interaction modes",
                        "Task or extension support",
                    ],
                    "allow_not_supported": ["Task or extension support"],
                },
            }

            supported_server_variants: list[str] = []
            for variant_name, specification in variants.items():
                section = mcp.section(specification["heading"])
                if section is None:
                    errors.append(
                        "Selected profile 'mcp-enabled' requires "
                        f"'{specification['heading']}' in MCP_INTERFACE.md."
                    )
                    continue

                interface_support = support_token(
                    mcp.field("Supported", section=section)
                )
                if interface_support not in {"YES", "NO"}:
                    errors.append(
                        f"MCP interface '{variant_name}' must set 'Supported:' "
                        "to YES or NO in MCP_INTERFACE.md."
                    )
                    continue

                if runtime is not None:
                    runtime_section = runtime.section(
                        specification["runtime_heading"]
                    )
                    runtime_support = support_token(
                        runtime.table_value(
                            "Supported", section=runtime_section
                        )
                    )
                    if runtime_support not in {"YES", "NO"}:
                        errors.append(
                            f"MCP variant '{variant_name}' must set Supported "
                            "to YES or NO in RUNTIME.md."
                        )
                    if (
                        runtime_support in {"YES", "NO"}
                        and interface_support != runtime_support
                    ):
                        errors.append(
                            f"MCP variant '{variant_name}' has inconsistent "
                            "Supported values between RUNTIME.md and "
                            "MCP_INTERFACE.md."
                        )

                if interface_support != "YES":
                    continue

                if variant_name != "bundled MCP client":
                    supported_server_variants.append(variant_name)
                if re.search(
                    r"\b(?:TODO|UNSELECTED)\b",
                    section,
                    re.IGNORECASE,
                ):
                    errors.append(
                        f"Supported MCP interface '{variant_name}' must not "
                        "retain TODO or UNSELECTED fields in MCP_INTERFACE.md."
                    )

                for label in specification["mandatory"]:
                    value = mcp.field(label, section=section)
                    valid = (
                        ValuePolicy.resolved(value)
                        if label in specification["allow_not_supported"]
                        else ValuePolicy.concrete(value)
                    )
                    if not valid:
                        errors.append(
                            f"Supported MCP interface '{variant_name}' requires "
                            f"a concrete '{label}:' value in MCP_INTERFACE.md."
                        )

            if not supported_server_variants:
                errors.append(
                    "Selected profile 'mcp-enabled' requires at least one "
                    "supported MCP server variant in MCP_INTERFACE.md."
                )

    if errors:
        for error in dict.fromkeys(errors):
            print(error, file=sys.stderr)
        return 1

    print("Extended Agent Skill profile contracts are valid.")
    return 0


if __name__ == "__main__":
    raise SystemExit(run())
